From dd123996c7b5f0c0e1f463725437c601c918495b Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Fri, 20 Dec 2019 14:35:21 +0200
Subject: [PATCH 30/30] sdk_dpaa: ceetm: save the root class pointer

commit d39cc9ffcbe5638b50f5f45698eb87a6c3a96eb3 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

When determining the parent prio qdisc of a new wbfs qdisc, save the
pointer to the root class as well.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 .../net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c  | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
index d3b97f07800b..00078010a6ac 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_ceetm.c
@@ -908,7 +908,7 @@ static int ceetm_init_prio(struct Qdisc *sch, struct ceetm_qdisc *priv,
 static int ceetm_init_wbfs(struct Qdisc *sch, struct ceetm_qdisc *priv,
 			   struct tc_ceetm_qopt *qopt)
 {
-	struct ceetm_class *parent_cl, *child_cl, *root_cl;
+	struct ceetm_class *parent_cl, *child_cl, *tmp_cl, *root_cl = NULL;
 	struct Qdisc *root_qdisc, *parent_qdisc = NULL;
 	struct net_device *dev = qdisc_dev(sch);
 	unsigned int i, id, prio_a, prio_b;
@@ -934,12 +934,13 @@ static int ceetm_init_wbfs(struct Qdisc *sch, struct ceetm_qdisc *priv,
 
 	/* Obtain the root ceetm class and the parent prio ceetm qdisc */
 	for (i = 0; i < root_priv->clhash.hashsize; i++) {
-		hlist_for_each_entry(root_cl, &root_priv->clhash.hash[i],
+		hlist_for_each_entry(tmp_cl, &root_priv->clhash.hash[i],
 				     common.hnode) {
-			if (root_cl->root.child &&
-			    (TC_H_MAJ(root_cl->root.child->handle) ==
+			if (tmp_cl->root.child &&
+			    (TC_H_MAJ(tmp_cl->root.child->handle) ==
 			    TC_H_MAJ(sch->parent))) {
-				parent_qdisc = root_cl->root.child;
+				parent_qdisc = tmp_cl->root.child;
+				root_cl = tmp_cl;
 				break;
 			}
 		}
-- 
2.17.1

